### ISSUE TITLE
Are there plans to extend this to other primitives?

### ISSUE THREAD

**Issue by BolunDai0216** [kevin-tracy/diffpills_jax#1]
Hi,

Thank you for open-sourcing this, `DifferentiableCollisions.jl` was awesome, but communication between Julia and Python has often been a bottleneck. 

I was wondering if you are planning to extend this Python version to other primitives and eventually have the same functionalities as `DifferentiableCollisions.jl`. I would love to help since my research has recently relied on `DifferentiableCollisions.jl` a lot.

**Comment by kevin-tracy** [comment]
Thanks for the note. 

I would love to do this but I just haven't had the time. In general I think there are a few options:

1. A direct port of `DifferentiableCollisions.jl` to C++ and a python wrapper with pybind11. This would be great for getting people to use it, but it would be stuck to the CPU. 
2. `DifferentiableCollisions.jl` but in torch or jax (to run on GPU).

Option 2 is easier but I haven't quite figured out a way to get things to play nice with jax's `vmap`. It doesn't seem like there is any obvious way to truly parallelize many collision computations unless they are all the same two primitives.  Since `DifferentiableCollisions.jl` is based on multiple dispatch / operator overloading, it's not very "pythonic". 

I'm open to any and all ideas/suggestions.

**Comment by BolunDai0216** [comment]
I think the second option is what would be exciting. At the moment, I don't have a clear picture of how this can be done, but I will look into it. 

I think you mentioned in the paper that `DifferentiableCollisions.jl` is based on `conelp` from `CVXOPT`, I'm not sure what other features you have added, but I can start to look at if solving for the scaling factor using a `conelp`-like solver can be batched using `jax`.

**Comment by BolunDai0216** [comment]
I noticed in `jax` you are directly using the `grad` function to compute the gradient instead of using the implicit function theorem. Do you think this would work also for the DifferentiableCollisions formulation?

**Comment by kevin-tracy** [comment]
Good catch, yes it does.  If the only gradient you want is that of the objective value with respect to the problem parameters, you can simply differentiate the Lagrangian at the primal-dual solution with respect to the problem parameters (no implicit function theorem or linear system solve required). I implemented this a while ago in `DifferentiableCollisions.jl` with the `proximity_gradient` function:

```julia
# return min scaling α and gradient of α wrt configurations 
α, dα_dstate = dc.proximity_gradient(P1, P2; verbose = false, pdip_tol = 1e-6)
```

This should be ~15-20% faster than using the implicit function theorem.
